update ds/hash (#39)
OI-wiki/i18n: docs/ds/hash.md

## Hash table

> [Hash table](https://en.wikipedia.org/wiki/Hash_table), also known as hash map, is a data structure that stores data in the form of "key-value". The so-called storage of data in the form of "key-value" means that any key corresponds uniquely to a certain location in memory. You only need to enter the searched value key to quickly find its corresponding value. The hash table can be understood as a kind of advanced array. The index of this array can be a large integer, floating point number, string or even a structure.

Reviewer: ```suggestion
[Hash table](https://en.wikipedia.org/wiki/Hash_table), also known as hash map, is a data structure that stores data in the form of "key-value". The so-called storage of data in the form of "key-value" means that any key corresponds uniquely to a certain location in memory. You only need to enter the searched value key to quickly find its corresponding value. The hash table can be understood as a kind of advanced array. The index of this array can be a large integer, a floating point number, a string, or even a structure.
```
to quickly find :thinking:
You can find corresponding val quickly with (just) the key of that value only?

> [Hash table](https://en.wikipedia.org/wiki/Hash_table), also known as hash map, is a data structure that stores data in the form of "key-value". The so-called storage of data in the form of "key-value" means that any key corresponds uniquely to a certain location in memory. You only need to enter the searched value key to quickly find its corresponding value. The hash table can be understood as a kind of advanced array. The index of this array can be a large integer, a floating point number, a string, or even a structure.

Reviewer: The so-called storage of data in the form of 这里其实可以用 which 直接连起来的
searched value key :thinking: 考虑更好的翻译

In OI, the most common case is using key as an integer. When the range of the key is relatively small, you can directly use it as the index of the array, but when the range is relatively large, such as using an integer in the range of $10^9$ as the key, a hash table is needed. Generally, the key modulo a larger prime number is used as the index, that is, $f(x)=x \bmod M$ is used as the hash function. Another common situation is when the key is a string. In OI, the general case is that string is not directly used as the key, but the hash value of the string is calculated first, and then the hash value is inserted into the hash table as the key.

> After we calculate the index for the key, we can know where each value should be placed. Suppose we use array a to store data, and the hash function is f, then the key-value pair (key, value) should be placed on a[f(key)]. No matter what type of key is and how large the range is, f(key) is an integer within the acceptable range and can be used as the index of the array.

Reviewer: ```suggestion
After we calculate the index for the key, we can know where each value should be placed. Suppose we use array $a$ to store data, and the hash function is $f$, then the key-value pair $(\text{key}, \text{value})$ should be placed on $a[f(\text{key})]$. No matter what type of key is and how large the range is, $f(\text{key})$ is an integer within the acceptable range and can be used as the index of the array.
```
中文语境下这样数组 a 应该问题不大，英文下可能会带来迷惑
（自己的修改）
🤔

## Hash collision

> If the index calculated by the hash function is different for any key, then just put (key, value) in the corresponding position according to the index. But in fact, there are often two different keys, and their indexes calculated by the hash function are the same. At this time, we would be needing some methods to deal with collisions. In OI, the most commonly used one is the zipper method.

Reviewer: ```suggestion
If the index calculated by the hash function is different for any key, then just put (key, value) in the corresponding position according to the index. But in fact, there are often two different keys, but their indexes calculated by the hash function are the same. At this time, we would need some methods to deal with collisions. In OI, the most commonly used one is the zipper method.
```

### Zipper method

Zipper method can also be called open hashing.

> The zipper method is to open a linked list in each place where data is stored. If there are multiple keys with the same index, just put them all in the linked list at that location. When querying, you need to scan the entire linked list at the corresponding location, and compare whether the key is consistent with the query key for each data. If the range of the index is 1~M and the size of the hash table is N, then an insert/query needs to perform comparisons in expected time complexity of $O(\frac{N}{M})$ .

Reviewer: ”开一个“ create

### Closed hashing 

The closed hashing method stores all records directly in the hash table. If there is a collision, it will continue probing in a certain way.

> For example, linear probing method: if there is a conflict at `d`, check `d + 1`, `d+2` in turn...

Reviewer: ```suggestion
For example, linear probing method: if there is a conflict at `d`, check `d+1`, `d+2` in turn...
```

};
```

Here we will provide you with a encapsulated template. It can be used like a map and is shorter.

```cpp
struct hash_map {  // hash table template
  struct data {
    long long u;
    int v, nex;
  };                // forward star list: see reference below

>   data e[SZ << 1];  // SZ is const int, representing the size

Reviewer: ```suggestion
  data e[SZ << 1];  // SZ is a const int representing the size
```

};
```

> > [forward star list](https://zh.wikipedia.org/wiki/%E9%93%BE%E5%BC%8F%E5%89%8D%E5%90%91%E6%98%9F) is a data structure used to store graph. It was named by [Malash](https://malash.me/200910/linked-forward-star/) but the actual inventor of the algorithm remains unknown.

Reviewer: :+1:
